## Subject 1: MNIST Clustering

1. Download the MNIST data set using the code below.
2. Ignoring the label normally associated to the dataset, construct a clustering of the data. Your clustering should maximize the v-score measure relative to true data labels (
    Paper describing the measure: https://www.aclweb.org/anthology/D07-1043.pdf
    Implementation available in python: https://scikit learn.org/stable/modules/generated/sklearn.metrics.v_measure_score.html#sklearn.metrics.v_measure_score
). Note that failing to ignore the labels during training will void your score for this subject.

In [2]:
from sklearn.datasets import fetch_openml
# Load data from https://www.openml.org/d/554
X, y = fetch_openml('mnist_784', version=1, return_X_y=True)

#vrem sa vedem ce fel de elemnte contin X si y
print(X), print(y)

[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]
['5' '0' '4' ... '4' '5' '6']


(None, None)

In [3]:
#rescriem y pentru a fi cat mai clar care este coloana de label, care este ignorata deja,
#pentru ca inca de la inceput este separata de restul datelor
print(y), type(y)

['5' '0' '4' ... '4' '5' '6']


(None, numpy.ndarray)

In [4]:
#forma dataframe a datelor este: 
import pandas as pd
df= pd.DataFrame(X,y)
df

0    1    2    3    4    5    6    7    8    9    ...  774  775  776  777  \
5   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
0   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
4   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
1   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
9   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
..  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
2   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
3   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
4   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
5   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
6   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   

    778  779  780  781  782  783  
5   0.0  0.0  0.0  0.0  0.0  0.0  
0   0.0  0.0  0.0  0.0  0.0  0.0  
4   0.0  0.0  0.0  0.0  0.0  0.0  
1   0.0  0.0  0.0  0.0  0.0  0.0  
9   0.0  0.0  0.0  0.0  0.0  0.0  
..  ...  ...  ...  ...  ...  ...  
2   0.0  0.0  0.0  0.0  0.0  0.0  
3   0.0  0.0  0.0  0.0  0.0  0.0  
4   0.0  0.0  0.0  0.0  0.0  0.0  
5   0.0  0.0  0.0  0.0  0.0  0.0  
6   0.0  0.0  0.0  0.0  0.0  0.0  

[70000 rows x 784 columns]

In [5]:
df.head()

0    1    2    3    4    5    6    7    8    9    ...  774  775  776  777  \
5  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
9  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   

   778  779  780  781  782  783  
5  0.0  0.0  0.0  0.0  0.0  0.0  
0  0.0  0.0  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  0.0  0.0  
9  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 784 columns]

In [6]:
import matplotlib.pyplot as plt
from matplotlib.cm import binary

some_digit_image = X[1].reshape(28, 28)
plt.imshow(some_digit_image, cmap=binary, interpolation="nearest")
plt.axis('off')
plt.show()

<Figure size 640x480 with 1 Axes>

In [7]:
# normalizam datele prin impartirea prin 255
X_norm=X/255.

In [13]:
#facem KMeans pentru k=10, pentru cele 10 caractere 
from sklearn.cluster import KMeans
import numpy as np

kmeans = KMeans(n_clusters=10, random_state=0,max_iter=500, n_init=15).fit(X_norm)

In [14]:
#determinam v-score pentru k=10
from sklearn.metrics.cluster import v_measure_score
v_measure_score(y, kmeans.labels_)

0.49974303091937844

In [ ]:
#observam ca nu e o valoare foarte buna si mai incercam cu mai multe posibilitati pentru k

In [16]:
kmeans = KMeans(n_clusters=2, random_state=0,max_iter=500, n_init=15).fit(X_norm)
v_measure_score(y, kmeans.labels_)

0.1642698162944342

In [ ]:
#este evident ca valoarea lui v-score pentru 2 nuclee este mult prea mica
#prima, pentru 10 nuclee, e mult mai buna, dar mai incerc si valori mai mari, si valori putin mai mici

In [24]:
# importam librariile necesare
from sklearn.cluster import KMeans
import numpy as np
from sklearn.metrics.cluster import v_measure_score
v_score=[]

# in k means clustering se vor testa intre  6 si 11 clustere
for j in range(6,12):
    # fitting k-means clustering
    kmeans = KMeans(n_clusters=j, random_state=0).fit(X_norm)
    # appending the score 
    v_score.append(v_measure_score(y, kmeans.labels_))

In [25]:
# facem o lista care asociaza numarului de nuclee v-scorul sau
print(dict(zip(range(6,12),v_score)))

{6: 0.45175031159301016, 7: 0.47041202409755334, 8: 0.5063185078368722, 9: 0.49246848728550235, 10: 0.49974303091937844, 11: 0.4971195507583373}


The best number of clusters comes out to be 8 with a v_score of 0.506 , however, the number of clusters = 10 is also very close. 

## Subject 2: Text NewsGroup Classification
1. Download the newsgroups data set using the code below. 
2. Construct a text classifier that predicts the target variable (newsgroups.target) from the input data (newsgroups.data).
3. We will evaluate your classifier against a hold-out data set, so be sure to construct a classification function that can receive a single string.

In [12]:
# importam newsgroups din sklearn
from sklearn.datasets import fetch_20newsgroups
newsgroups = fetch_20newsgroups(subset='train')

In [13]:
# importam countvectorizer si Naive Bayes libraries
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB

# realizam o instanta pentru count vectorizer
count_vect = CountVectorizer(strip_accents='ascii', token_pattern=u'(?ui)\\b\\w*[a-z]+\\w*\\b', 
                     lowercase=True, stop_words='english')

# transformam datele intr-un vector de counts 
X_transformed = count_vect.fit_transform(newsgroups.data)


# fitting Naive Bayes on the count data
mod = MultinomialNB()
mod.fit(X_transformed, newsgroups.target)
predictions = mod.predict(X_transformed)

In [14]:
#evolutia rezultatelor prezise in rsaport cu labelurile reale
from sklearn.metrics import classification_report
print(classification_report(y_true=newsgroups.target,y_pred=predictions))

              precision    recall  f1-score   support

           0       0.95      0.99      0.97       480
           1       0.84      0.97      0.90       584
           2       0.97      0.24      0.39       591
           3       0.72      0.97      0.83       590
           4       0.93      0.98      0.96       578
           5       0.84      0.98      0.91       593
           6       0.96      0.93      0.95       585
           7       0.97      0.98      0.98       594
           8       1.00      0.99      0.99       598
           9       0.99      0.99      0.99       597
          10       0.98      0.99      0.99       600
          11       0.96      0.99      0.98       595
          12       0.96      0.97      0.97       591
          13       0.99      0.99      0.99       594
          14       0.98      1.00      0.99       593
          15       0.98      0.99      0.99       599
          16       0.97      0.99      0.98       546
          17       0.99    

In [15]:
# please enter the texts in a list as newsgroup.data
#functia urmatoare face predictie pe baza clasificatorului potrivit mai sus

def predict_val(text):
    
    X_trans = count_vect.transform(text)
    
    predictions = mod.predict(X_trans)
    return(predictions)

In [16]:
# testarea functiei pentru cateva valori
predict_val([newsgroups.data[0]])

array([7])

In [17]:
predict_val([newsgroups.data[7]])

array([3])

In [18]:
predict_val([newsgroups.data[700]])

array([1])